platform/broadcom/sonic-platform-modules-cel/silverstone/modules/i2c-ocores.c

```diff
@@ -0,0 +1,846 @@
+// SPDX-License-Identifier: GPL-2.0
```

Comment:
https://elixir.bootlin.com/linux/v4.1/source/drivers/i2c/busses/i2c-ocores.c

why not use linux kernel source code?

Reply:
Hello,

The reason we keep it as a separate are:
1. The original i2c-ocores in sonic kernel version 4.9.x is not support polling mode. We use the codes based on kernel 5.1 to support this feature.
2. We add a workaround to fix read error on some SFP loopback module.

If only the first reason, we can backport the feature from kernel 5.1 to current sonic-linux-kernel, but combined with the 2nd reason, we decide to maintain the codes as a platform-specific.

Reply:
why do we need this workaround? We add a workaround to fix read error on some SFP loopback module.

Reply:
In some SFP module case.
When reading EEPROM, the i2c core detects the arbitration lost(AL) error and stop the i2c transfer. The AL randomly occurs on the device.

To make the device readable, the limited retry is added to the ocore FSM. When AL detects it restarts the i2c transfer again until retry used up.

Reply:
this seems like a generic fix, can you patch the kernel in sonic-linux-kernel?

Reply:
Yes, I will create a PR to patch i2c-ocores.c in sonic-linux-kernel to support polling mode.
For the workaround, we found that it was a bug in the FPGA firmware and fixed with new firmware, so it will not include in this patch.

Reply:
thanks. I will review your sonic-linux-kernel patch and merge this pr after kernel patch has been fixed.

Reply:
The kernel patch has been introduced here https://github.com/Azure/sonic-linux-kernel/pull/135.

Reply:
The driver is updated to use kernel's i2c-ocores module.